### PR TITLE
bpo-30781: IDLE - use ttk widgets in configdialog

### PR DESCRIPTION
**Pull request by csabella**
First set of commits for converting constants to string literals.

https://bugs.python.org/issue30781

### CODE REVIEW

**Comment by mlouielu** [comment]
Maybe I'm not on the date, does this a new policy to changed tkinter constant to string?

**Comment by csabella** [comment]
On the last file that was converted to use ttk widgets, I also changed the constants to literal strings at the same time because Terry prefers the literal strings.  I was applying the same steps here.

**Comment by csabella** [comment]
The latest commit converts to ttk widgets.  Some comments about the changes:
1.  In order to show the color changes for the themes, I needed to use a Frame background.  The only way to do this in ttk is with a style, but I tried to minimize the impact of adding Style since we don't want to convert to using that yet.
2.  The `Checkbutton` no longer has an `indicatoron` option, so they look a lot different on the extensions tab.
3.  I didn't convert `Scale` because the ttk version didn't have a `tickinterval`.  I believe this is being changed anyway, so I left it as is for now.

**Comment by terryjreedy** [comment]
I am thinking that having NORMAL and DISABLED stand out is ok.  Don't change though yet.
The change of from .config to subscripting created conflicts.  We may introduce others before we get to this.

**Comment by csabella** [comment]
Yes, I had been keeping up with resolving the conflicts, but haven't done it in a few days, since the restructuring to the classes got the momentum.  Once you're ready for this, I'll just do it all then.

File: Lib/idlelib/idle_test/test_configdialog.py [hunk 1]
```diff
@@ -272,7 +272,7 @@ def test_load_theme_cfg(self):
         # builtinlist sets variable builtin_name to the CurrentTheme default.
         eq(d.builtin_name.get(), 'IDLE Classic')
         eq(d.custom_name.get(), '- no custom themes -')
-        eq(d.custom_theme_on['state'], DISABLED)
+        eq(d.custom_theme_on.state(), (DISABLED,))
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_configdialog.py]: Make this 'disabled' to be consistent with '!disabled'.

File: Lib/idlelib/configdialog.py [hunk 2]
```diff
@@ -9,13 +9,10 @@
 Refer to comments in EditorWindow autoindent code for details.
 
 """
-from tkinter import (Toplevel, Frame, LabelFrame, Listbox, Label, Button,
-                     Entry, Text, Scale, Radiobutton, Checkbutton, Canvas,
-                     StringVar, BooleanVar, IntVar, TRUE, FALSE,
-                     TOP, BOTTOM, RIGHT, LEFT, SOLID, GROOVE, NORMAL, DISABLED,
-                     NONE, BOTH, X, Y, W, E, EW, NS, NSEW, NW,
-                     HORIZONTAL, VERTICAL, ANCHOR, ACTIVE, END)
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Leave these imports.

File: Lib/idlelib/configdialog.py [hunk 3]
```diff
@@ -814,19 +809,21 @@ def create_page_highlight(self):
...
+            text.insert('end', texttag[0], texttag[1])
         for element in self.theme_elements:
             def tem(event, elem=element):
                 # event.widget.winfo_top_level().highlight_target.set(elem)
                 self.highlight_target.set(elem)
             text.tag_bind(
                     self.theme_elements[element][0], '<ButtonPress-1>', tem)
-        text['state'] = DISABLED
-        self.frame_color_set = Frame(frame_custom, relief=SOLID, borderwidth=1)
+        text['state'] = 'disabled'
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:824]: Keep

File: Lib/idlelib/configdialog.py [hunk 4]
```diff
@@ -913,7 +910,7 @@ def load_theme_cfg(self):
             item_list = idleConf.GetSectionList('user', 'highlight')
             item_list.sort()
             if not item_list:
-                self.custom_theme_on['state'] = DISABLED
+                self.custom_theme_on.state(('disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:918]: Keep.

File: Lib/idlelib/configdialog.py [hunk 5]
```diff
@@ -1004,14 +999,14 @@ def set_theme_type(self):
...
+            self.button_delete_custom.state(('disabled',))
         else:
-            self.builtinlist['state'] = DISABLED
-            self.custom_theme_on['state'] = NORMAL
-            self.customlist['state'] = NORMAL
-            self.button_delete_custom['state'] = NORMAL
+            self.builtinlist['state'] = 'disabled'
+            self.custom_theme_on.state(('!disabled',))
+            self.customlist['state'] = 'normal'
+            self.button_delete_custom.state(('!disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1014]: Keep block.

File: Lib/idlelib/configdialog.py [hunk 6]
```diff
@@ -1139,12 +1135,12 @@ def set_highlight_target(self):
...
-            self.fg_on['state'] = DISABLED
-            self.bg_on['state'] = DISABLED
+            self.fg_on.state(('disabled',))
+            self.bg_on.state(('disabled',))
             self.fg_bg_toggle.set(1)
         else:  # Both fg and bg can be set.
-            self.fg_on['state'] = NORMAL
-            self.bg_on['state'] = NORMAL
+            self.fg_on.state(('!disabled',))
+            self.bg_on.state(('!disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1148]: Keep block.

File: Lib/idlelib/configdialog.py [hunk 7]
```diff
@@ -1260,7 +1256,7 @@ def delete_custom(self):
         item_list = idleConf.GetSectionList('user', 'highlight')
         item_list.sort()
         if not item_list:
-            self.custom_theme_on['state'] = DISABLED
+            self.custom_theme_on.state(('disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1264]: Keep.

File: Lib/idlelib/configdialog.py [hunk 8]
```diff
@@ -1395,9 +1391,9 @@ def create_page_keys(self):
         self.bindingslist['xscrollcommand'] = scroll_target_x.set
         self.button_new_keys = Button(
                 frame_custom, text='Get New Keys for Selection',
-                command=self.get_new_keys, state=DISABLED)
+                command=self.get_new_keys, state='disabled')
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Keep.

File: Lib/idlelib/configdialog.py [hunk 9]
```diff
@@ -1457,7 +1453,7 @@ def load_key_cfg(self):
             item_list = idleConf.GetSectionList('user', 'keys')
             item_list.sort()
             if not item_list:
-                self.custom_keyset_on['state'] = DISABLED
+                self.custom_keyset_on.state(('disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1461]: Keep.

File: Lib/idlelib/configdialog.py [hunk 10]
```diff
@@ -1526,14 +1520,14 @@ def var_changed_keybinding(self, *params):
...
+            self.button_delete_custom_keys.state(('disabled',))
         else:
-            self.builtinlist['state'] = DISABLED
-            self.custom_keyset_on['state'] = NORMAL
-            self.customlist['state'] = NORMAL
-            self.button_delete_custom_keys['state'] = NORMAL
+            self.builtinlist['state'] = 'disabled'
+            self.custom_keyset_on.state(('!disabled',))
+            self.customlist['state'] = 'normal'
+            self.button_delete_custom_keys.state(('!disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1535]: Keep block.

File: Lib/idlelib/configdialog.py [hunk 11]
```diff
@@ -1595,7 +1589,7 @@ def save_as_new_key_set(self):
 
     def on_bindingslist_select(self, event):
         "Activate button to assign new keys to selected action."
-        self.button_new_keys['state'] = NORMAL
+        self.button_new_keys.state(('!disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1597]: Keep.

File: Lib/idlelib/configdialog.py [hunk 12]
```diff
@@ -1689,7 +1683,7 @@ def delete_custom_keys(self):
         item_list = idleConf.GetSectionList('user', 'keys')
         item_list.sort()
         if not item_list:
-            self.custom_keyset_on['state'] = DISABLED
+            self.custom_keyset_on.state(('disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1691]: Keep.

File: Lib/idlelib/configdialog.py [hunk 13]
```diff
@@ -1803,49 +1797,49 @@ def create_page_general(self):
...
                 frame_helplist, height=5, takefocus=True,
-                exportselection=FALSE)
+                exportselection=False)
         scroll_helplist = Scrollbar(frame_helplist)
         scroll_helplist['command'] = self.helplist.yview
         self.helplist['yscrollcommand'] = scroll_helplist.set
         self.helplist.bind('<ButtonRelease-1>', self.help_source_selected)
         self.button_helplist_edit = Button(
-                frame_helplist_buttons, text='Edit', state=DISABLED,
+                frame_helplist_buttons, text='Edit', state='disabled',
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1811]: Keep.

File: Lib/idlelib/configdialog.py [hunk 14]
```diff
@@ -1803,49 +1797,49 @@ def create_page_general(self):
...
                 width=8, command=self.helplist_item_edit)
         self.button_helplist_add = Button(
                 frame_helplist_buttons, text='Add',
                 width=8, command=self.helplist_item_add)
         self.button_helplist_remove = Button(
-                frame_helplist_buttons, text='Remove', state=DISABLED,
+                frame_helplist_buttons, text='Remove', state='disabled',
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1817]: Keep.

File: Lib/idlelib/configdialog.py [hunk 15]
```diff
@@ -1874,15 +1868,15 @@ def help_source_selected(self, event):
...
             if self.helplist.curselection():  # There currently is a selection.
-                self.button_helplist_edit['state'] = NORMAL
-                self.button_helplist_remove['state'] = NORMAL
+                self.button_helplist_edit.state(('!disabled',))
+                self.button_helplist_remove.state(('!disabled',))
             else:  # There currently is not a selection.
-                self.button_helplist_edit['state'] = DISABLED
-                self.button_helplist_remove['state'] = DISABLED
+                self.button_helplist_edit.state(('disabled',))
+                self.button_helplist_remove.state(('disabled',))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:1884]: Keep.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
For the ttk issue, I want a minimal patch that only does what is needed to make the switch, or to be consistent about state ('disabled' instead of DISABLED with '!disabled', and 'normal').  I noted most but perhaps not all of these.

   I believe everything other that the one note in the tests is needed.  For configdialog, 80% is extraneous, so I presume it would be easier to start fresh.

I definitely want to switch to True, False, and widget['option'] = value, but in a separate style issue, at least after disposing of existing patches.

**Comment by csabella** [comment]
I didn't expect the Spanish Inquisition!

File: Lib/idlelib/configdialog.py [hunk 16]
```diff
@@ -9,13 +9,13 @@
...
                      StringVar, BooleanVar, IntVar, TRUE, FALSE,
                      TOP, BOTTOM, RIGHT, LEFT, SOLID, GROOVE, NORMAL, DISABLED,
                      NONE, BOTH, X, Y, W, E, EW, NS, NSEW, NW,
                      HORIZONTAL, VERTICAL, ANCHOR, ACTIVE, END)
-from tkinter.ttk import Notebook, Scrollbar
+from tkinter.ttk import (Button, Checkbutton, Entry, Frame, Label, LabelFrame,
+                         Notebook, Radiobutton, Scrollbar, Style)
 import tkinter.colorchooser as tkColorChooser
 import tkinter.font as tkFont
 import tkinter.messagebox as tkMessageBox
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py:21]: I want to change these in a subsequent 'style' patch, either with other style changes for this file or a separate patch for all files.  Probably
```
from tkinter import colorchooser
from tkinter import messagebox
import tkinter.font as tkfont
```
I think `tkfont` may be necessary to not conflict with local `font` names.  Will check with grep first.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Pending a 'human test', looks good.

In title of existing news blurb,  Misc/NEWS.d/next/IDLE/2017-07-28-18-59-06.bpo-30781.ud5m18.rst, change 'Notebook' to 'widgets'.  Add 'Patches by Terry Jan Reedy and Cheryl Sabella.

**Comment by csabella** [comment]
I didn't expect the Spanish Inquisition!

**Review by terryjreedy — APPROVED** [review]

**Pull request merged**